design-proposals/dynamic-primary-pod-nic-name.md

```diff
@@ -0,0 +1,466 @@
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  name: kubevirt
+  namespace: kubevirt
+spec:
+  configuration:
+    network:
+      preferredPodPrimaryInterfaceName: special-nic
+```
+
+virt-handler and virt-launcher will be adjusted to prefer this value when trying to bind the primary network interface.
+They will both fallback to `eth0` in case they cannot find the preferred interface name.
+
+#### Pros
+
+- Does not require Multus to work
+
+#### Cons
+
+- Requires manual configuration
+- Error prone
+- Has cluster-wide granularity
+- Requires a change to the communication protocol between virt-handler and virt-launcher to pass the cluster-wide
+  setting to virt-launcher
+
+### Specifying the primary pod NIC name in advance
+
+When creating a VM with secondary networks, KubeVirt specifies the pod interfaces’ names for all secondary interfaces.
+It is currently not possible to specify the primary interface’s name when templating the virt-launcher pod, since there
+is no K8s API that enables it.
+
+The primary pod interface name is set by the container runtime (for example cri-o), before invoking the default CNI.
+
+#### Pros
+
+#### Cons
+
+- There is no K8s API to name the pod's primary NIC
+- There is no Multus API or config to name the pod's primary NIC
+
+## Network Binding Plugins
+
+A virtual machine's NIC could be bound to the pod's primary NIC using
+a [network binding plugin](https://kubevirt.io/user-guide/network/network_binding_plugins/).
+
+### CNI
+
+KubeVirt's VMI controller is responsible for specifying the proper [NetworkSelectionElement](https://github.com/k8snetworkplumbingwg/network-attachment-definition-client/blob/506cfdac925790adf2f56f27740d2e87eaf2c83c/pkg/apis/k8s.cni.cncf.io/v1/types.go#L140)
+objects that will cause Multus to invoke a network binding plugin's CNI.
+
+Since this specification happens before the pod is created, it is not possible for KubeVirt to know in advance what will
+the pod's primary NIC's name be.
+
+Binding plugins' authors could add a configuration field to specify a list of possible primary interface names:
+
+```json
+{
+  ...
+  "potentialPrimaryInterfaceNames": "custom-iface,eth0",
+  ...
+}
+```
+These will be specified under the plugin's NetworkAttachmentDefinition object.
+
+The binding plugin's CNI could look up these names in order.
+
+### Sidecar
+
+Network binding plugins could use
+a [sidecar container](https://github.com/kubevirt/kubevirt/blob/6aba1bc426d8b2f93270ba649ca33b9bbd925729/docs/network/network-binding-plugin.md?plain=1#L177)
+in order to:
+
+- Mutate the domain xml that is passed to libvirt's daemon (virtqemud).
+- Run arbitrary code in the context of the virt-launcher pod.
+
+Since protocol between the virt-launcher and the sidecar provides the VirtualMachineInstanceObject, the sidecar could
+infer what is the pod's primary interface by examining it.
+
+In that case, a change to the network binding plugins' developer guide will be required.
```

Comment:
Please make sure to also describe how we assure (or what we assume) the call to the sidecar is with an updated information. E.g. that it is not called before the relevant data is appearing in the spec/status.